clippy_lints/src/methods/string_lit_chars_any.rs

```diff
@@ -0,0 +1,80 @@
+use clippy_utils::{
+    diagnostics::span_lint_and_then,
+    is_from_proc_macro, is_trait_method,
+    msrvs::{Msrv, MATCHES_MACRO},
+    path_to_local,
+    source::snippet_opt,
+};
+use itertools::Itertools;
+use rustc_ast::LitKind;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, Expr, ExprKind, HirId, Param, PatKind};
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::STRING_LIT_CHARS_ANY;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    recv: &Expr<'_>,
+    param: &'tcx Param<'tcx>,
+    body: &Expr<'_>,
+    msrv: &Msrv,
+) {
+    if !msrv.meets(MATCHES_MACRO) || !is_trait_method(cx, expr, sym::Iterator) {
+        return;
+    };
+
+    // All bindings introduced by `param`
+    let mut bindings = FxHashSet::default();
+
+    param.pat.walk_always(|pat| {
+        if let PatKind::Binding(_, hir_id, _, _) = pat.kind {
+            bindings.insert(hir_id);
+        }
+    });
+
+    if let ExprKind::Lit(lit_kind) = recv.kind
+        && let LitKind::Str(val, _) = lit_kind.node
+        && let ExprKind::Binary(kind, lhs, rhs) = body.kind
+        && let BinOpKind::Eq = kind.node
+        && let Some(scrutinee) = lhs_rhs_are_param_and_scrutinee(&bindings, lhs, rhs)
+            .or_else(|| lhs_rhs_are_param_and_scrutinee(&bindings, rhs, lhs))
```

Comment:
Once the `HashSet` is changed to a single `HirId` this can be simplified to:
```
&& let Some(lhs) = path_to_local(lhs)
&& let Some(rhs) = path_to_local(rhs)
&& (lhs == arg) != (rhs == arg)
```